### ISSUE TITLE
[RFC] New Unform API

### ISSUE THREAD

**Issue by diego3g** [Rocketseat/unform#118]
## Main problem

Unform was released some months ago and we're receiving a lot of feature suggestions that can't be implemented with the current Unform API.

Example: If we want to set an input value programatically it's not possible. If we want to get/set an error message programatically it's not possible. If we want to use any form data outside the onSubmit function, it's not possible.

Due to this, i was studying ways to keep unform simple and fast but give developer access to simple functions like getting/setting an input value, getting/settings error messages, etc.

## Proposed API

After many tests i got something like this that i want to share with you guys:

```jsx
import React from 'react';
import axios from 'axios';
import * as Yup from 'yup';
import { useForm, Form, Input } from '@rocketseat/unform';

export default function App() {
  const [{ setValue }, formRef] = useForm({
    initialData: { name: 'Diego' },
    schema: Yup.object().shape({
      name: Yup.string().required(),
    })
  });

  async function handleCepChange(e) {
    const cep = e.target.value;

    const response = await axios.get(`https://viacep.com.br/ws/${cep}/json`);

    setValue('street', response.data.logradouro);
  }

  async handleSubmit(data) {

  }

  return (
    <Form onSubmit={handleSubmit} ref={formRef}>
      <Input name="cep" onChange={handleCepChange} />
      <Input name="street" />
    </Form>
  );
}
```

## Main points here:

- We are using ref in our form to access all it properties and methods;
- We do not listen to input changes from our form but from input directly so we still keep not controlling inputs;
- We are moving initialData, schema and other form properties (should we?), maybe we could keep this inside `<Form />` tag.

## RFC

We need suggestions and discussions about this new proposed API so every comment here is welcome ❤️

**Comment by italomlp** [comment]
> We are moving initialData, schema and other form properties (should we?), maybe we could keep this inside `<Form />` tag

I think that initially, while we migrating, we should keep both approaches (if possible, of course), and deprecate the older way, to remove in next version.

And, I think that we can pass the `onSubmit` to the side of `initialValues` place too, keeping all together (again, if possible, I didn't get the code, so I don't know if it's worth it). So, in the `<Form />` tag, we only pass the ref, and all form things will be on the hook `useForm`, like so:

```jsx
// ...
const [{ setValue }, formRef] = useForm({
  initialData: { name: 'Diego' },
  schema: Yup.object().shape({
    name: Yup.string().required(),
  }),
  onSubmit: handleSubmit
});

// ...
return (
  <Form ref={formRef}>
    <Input name="cep" onChange={handleCepChange} />
    <Input name="street" />
  </Form>
);
```

Finally, I think it's a great and pretty API and is very welcome. +1 for it.

**Comment by diego3g** [comment]
@italomlp This approach will help us with React Native version that you were developing. Something like this:
```jsx
const [{ onSubmit }, formRef] = useForm({
  onSubmit: handleSubmit
});

return (
  <Form ref={formRef}>
    <Input name="name" />
    <Button onPress={onSubmit}>Save</Button>
  </Form>
);
```

**Comment by italomlp** [comment]
@diego3g exactly! Do you already start to implement something or are waiting for more comments?

**Comment by diego3g** [comment]
@italomlp I'm giving some time so we can have a little bit more discussion over it and then i'll start implementing it, but i already did a test and it's possible to achieve what we're trying to todo with this API.

**Comment by italomlp** [comment]
@diego3g great! I will follow this issue and comments and if you need help to implement it, you can count on me. Also, I think that I'll wait for this API be done to end the implementation of React Native support.

**Comment by lookfirst** [comment]
`setValue('street'...)` is hard to make typesafe. It would be nice to connect the two.

Supporting async behavior is great.

**Comment by diego3g** [comment]
Unform released version 2.0 🎉

New Unform API is now available, check the docs do start using it: https://unform.dev
